File: tests/analytic_precision/supplemental/savefe_extended_normalization_contract.py
"""Exact WLS and reconstruction across categorical and heterogeneous saved FEs."""
import argparse
import importlib.util
import json
import os
from pathlib import Path

import numpy as np
from exact_wls_reference import fit

parser=argparse.ArgumentParser(description=__doc__)
parser.add_argument('--module',type=Path,required=True)
parser.add_argument('--backend',choices=('cpu','cuda'),required=True)
parser.add_argument('--out',type=Path,required=True)
args=parser.parse_args()
os.environ.update(XHDFE_GPU_BACKEND=args.backend,XHDFE_ABSORPTION_CACHE_MODE='off',XHDFE_MOBILITY_MODE='off')
spec=importlib.util.spec_from_file_location('py_hdfe_v11',args.module)
cpp=importlib.util.module_from_spec(spec);spec.loader.exec_module(cpp)
i=np.arange(192);worker=i//48;firm=(i//16)%3;year=(i//8)%2
x=(2*(i%2)-1)+.5*worker;z=2*((i//2)%2)-1;noise=.125*(2*((i//4)%2)-1)
records=[]
for family in ('one_fe','two_fe','three_fe','heterogeneous','pure_plus_fe','pure_only'):
    fes=[worker] if family in ('one_fe','pure_only') else [worker,firm]
    if family=='three_fe':fes.append(year)
    mixed=family in ('heterogeneous','pure_plus_fe','pure_only')
    include_worker=family not in ('pure_plus_fe','pure_only')
    columns=[(worker==g).astype(float) for g in range(4)] if include_worker else []
    if len(fes)>1:columns.extend((firm==g).astype(float) for g in range(1 if include_worker else 0,3))
    if family=='three_fe':columns.append((year==1).astype(float))
    if mixed:columns.extend((worker==g)*z for g in range(4))
    dummies=np.column_stack(columns)
    slopes=[(0,z,include_worker)] if mixed else None
    y=3+.75*x+noise+(2*worker if include_worker else 0)
    if len(fes)>1:y=y-firm
    if family=='three_fe':y=y+.5*year
    if mixed:y=y+(1.+worker)*z
    for weighted in (False,True):
        weights=1.+worker if weighted else np.ones(len(i))
        for intercept in (False,True):
            design=dummies
            if family=='pure_only' and intercept:design=np.column_stack((design,np.ones(len(i))))
            exact=fit(np.column_stack((design,x)),y,weights)
            reference=np.array([float(v) for v in exact['residuals']])
            for style in ('component','reghdfe'):
                os.environ['XHDFE_FE_NORMALIZE']=style
                for retain in (False,True):
                    model=cpp.HdfeRegressor(num_threads=2,max_iter=1000,tol=1e-8,
                        drop_singletons=False,fit_intercept=intercept,retain_fes=retain)
                    row=dict(family=family,weighted=weighted,style=style,intercept=intercept,retain=retain)
                    try:model.fit(y,x[:,None],fes=fes,slopes=slopes,weights=weights if weighted else None)
                    except RuntimeError as error:row.update(verdict='FAIL',error=str(error))
                    else:
                        beta=float(model.coef_[0]);residual=np.asarray(model.residuals_)
                        row.update(beta_error=abs(beta-float(exact['coefficients'][-1])),
                            residual_error=float(np.max(np.abs(residual-reference))),
                            variance_relative_error=abs(float(model.covariance_[0,0])/float(exact['variance'])-1))
                        if retain:
                            predicted=beta*x+sum(np.asarray(v) for v in model.fe_effects_)
                            if intercept:predicted+=float(model.coef_[-1])
                            row['reconstruction_error']=float(np.max(np.abs(y-predicted-reference)))
                        passed=all(value<=1e-8 for key,value in row.items() if key.endswith('error'))
                        passed=passed and model.converged_ and model.precision_certified_ and model.fe_recovery_converged_
                        row['verdict']='PASS' if passed and bool(model.gpu_used_)==(args.backend=='cuda') else 'FAIL'
                    records.append(row)
counts={verdict:sum(row['verdict']==verdict for row in records) for verdict in ('PASS','FAIL')}
with args.out.open('x') as handle:json.dump(dict(backend=args.backend,counts=counts,rows=records),handle,indent=2)
print(json.dumps(counts));raise SystemExit(counts['FAIL']!=0)

File: tests/analytic_precision/supplemental/savefe_extended_normalization_contract.R
args <- commandArgs(trailingOnly=TRUE)
stopifnot(length(args)==6L)
xhdfe_library <- normalizePath(args[1],mustWork=TRUE)
dependency_libraries <- strsplit(args[2],.Platform$path.sep,fixed=TRUE)[[1L]]
dependency_libraries <- dependency_libraries[nzchar(dependency_libraries)]
dependency_libraries <- vapply(dependency_libraries,normalizePath,character(1),mustWork=TRUE)
.libPaths(c(xhdfe_library,dependency_libraries,.libPaths()))
expected_package <- normalizePath(args[5],mustWork=TRUE)
expected_dll <- normalizePath(args[6],mustWork=TRUE)
stopifnot(identical(normalizePath(find.package("xhdfe",lib.loc=xhdfe_library),mustWork=TRUE),expected_package))
library(xhdfe,lib.loc=xhdfe_library)
library(jsonlite)
backend <- args[3]
loaded_dll <- normalizePath(getLoadedDLLs()[["xhdfe"]][["path"]],mustWork=TRUE)
stopifnot(identical(loaded_dll,expected_dll))
i <- 0:191
data <- data.frame(worker=i%/%48,firm=(i%/%16)%%3,year=(i%/%8)%%2,
                   x=2*(i%%2)-1+.5*(i%/%48),z=2*((i%/%2)%%2)-1,
                   noise=.125*(2*((i%/%4)%%2)-1))
rows <- list()
for (family in c("one_fe","two_fe","three_fe","heterogeneous","pure_plus_fe","pure_only")) {
  fes <- if (family %in% c("one_fe","pure_only")) "worker" else c("worker","firm")
  if (family=="three_fe") fes <- c(fes,"year")
  mixed <- family %in% c("heterogeneous","pure_plus_fe","pure_only")
  include_worker <- !family %in% c("pure_plus_fe","pure_only")
  basis <- list()
  if (include_worker) basis <- lapply(0:3,function(g) as.numeric(data$worker==g))
  if (length(fes)>1) basis <- c(basis,lapply((if (include_worker) 1 else 0):2,function(g) as.numeric(data$firm==g)))
  if (family=="three_fe") basis <- c(basis,list(as.numeric(data$year==1)))
  if (mixed) basis <- c(basis,lapply(0:3,function(g) (data$worker==g)*data$z))
  dummies <- do.call(cbind,basis)
  data$y <- 3+.75*data$x+data$noise+if (include_worker) 2*data$worker else 0
  if (length(fes)>1) data$y <- data$y-data$firm
  if (family=="three_fe") data$y <- data$y+.5*data$year
  if (mixed) data$y <- data$y+(1+data$worker)*data$z
  for (weighted in c(FALSE,TRUE)) for (intercept in c(FALSE,TRUE)) {
    w <- if (weighted) 1+data$worker else rep(1,nrow(data))
    design <- dummies
    if (family=="pure_only" && intercept) design <- cbind(design,1)
    design <- cbind(design,x=data$x)
    reference <- lm(data$y ~ 0+design,weights=w)
    stopifnot(reference$rank==ncol(design))
    reference_b <- tail(coef(reference),1)
    reference_v <- tail(diag(vcov(reference)),1)
    model_has_constant <- include_worker || length(fes)>1 || intercept
    centered_y <- if (model_has_constant) data$y-weighted.mean(data$y,w) else data$y
    adjusted_r2 <- 1-(sum(w*residuals(reference)^2)/(nrow(data)-reference$rank))/
      (sum(w*centered_y^2)/(nrow(data)-as.integer(model_has_constant)))
    for (style in c("component","reghdfe")) for (retain in c(FALSE,TRUE)) for (interface in c("matrix","formula")) {
      Sys.setenv(XHDFE_FE_NORMALIZE=style)
      fit_args <- list(threads=2,maxiter=1000,tol=1e-8,backend=backend,
                       save_fe=retain,drop_singletons=FALSE,vcov="unadjusted")
      if (weighted) fit_args$weights <- w
      if (interface=="matrix") {
        fit_args$y <- data$y;fit_args$X <- as.matrix(data["x"]);fit_args$fes <- data[fes]
        fit_args$fit_intercept <- intercept
        if (mixed) fit_args$slopes <- list(list(fe=1L,values=data$z,include_intercept=include_worker))
        estimator <- xhdfe_fit
      } else {
        terms <- fes
        if (mixed) terms[1] <- if (include_worker) "worker[z]" else "worker[[z]]"
        fit_args$fml <- as.formula(paste("y ~",if (!intercept) "0 + x" else "x","|",paste(terms,collapse=" + ")))
        fit_args$data <- data;estimator <- xhdfe
      }
      record <- list(family=family,weighted=weighted,intercept=intercept,style=style,retain=retain,interface=interface)
      result <- tryCatch({
        model <- do.call(estimator,fit_args)
        errors <- c(beta=unname(abs(model$coefficients[1]-reference_b)),
                    variance=unname(abs(model$vcov[1,1]/reference_v-1)),
                    residual=max(abs(model$residuals-residuals(reference))),
                    adjusted_r2=abs(model$r2_a-adjusted_r2))
        if (retain) {
          fitted <- model$coefficients[1]*data$x+Reduce(`+`,model$fe_effects)
          if (intercept) fitted <- fitted+tail(model$coefficients,1)
          errors <- c(errors,reconstruction=max(abs(data$y-fitted-residuals(reference))))
        }
        passed <- errors[["beta"]]<=1e-9 && errors[["variance"]]<=1e-8 &&
          all(errors[setdiff(names(errors),c("beta","variance"))]<=1e-8) &&
          model$converged && model$precision_certified &&
          model$fe_recovery_converged && isTRUE(model$gpu_used)==(backend=="cuda")
        list(verdict=if (passed) "PASS" else "FAIL",errors=as.list(errors))
      },error=function(e) list(verdict="FAIL",error=conditionMessage(e)))
      rows[[length(rows)+1L]] <- c(record,result)
    }
  }
}
passed <- sum(vapply(rows,function(r) r$verdict=="PASS",logical(1)))
write_json(list(passed=passed,total=length(rows),rows=rows,
                package=expected_package,dll=loaded_dll),args[4],auto_unbox=TRUE,pretty=TRUE,digits=NA)
cat("SAVEFE_R",backend,passed,"OF",length(rows),"\n")
quit(status=as.integer(passed!=length(rows)))
